test/bin/mdx-pp/gen_dune_rules.ml

```diff
@@ -0,0 +1,28 @@
+open Gen_rule_helpers
+
+(** Tests that the results of 'cd <dir> && ocaml-mdx pp test-case.md' is equal
+    to the content of the '.expected' file in <dir>. *)
+let pp_expect_action fmt dir =
+  Fmt.pf fmt
+    {|
+  (with-stdout-to %%{target}
+   (chdir %s
+    (run ocaml-mdx pp %a%s)))|}
+    dir.dir_name
+    pp_options dir.options
+    dir.test_file
+
+(** Tests that 'cd <dir> && ocaml-mdx pp [options] <file>' exits with a
+    failing code and that its output is equal to the content of the
+    '.expected' file in <dir>. *)
+let pp_failure_action fmt dir =
+  Fmt.pf fmt
+    {|
+  (with-outputs-to %%{target}
+   (chdir %s
+    (system "! ocaml-mdx pp %a%s)))|}
```

Comment:
Repeating this strings, indenting and passing more arguments to pf could be avoided by having utils that print the `(with-outputs-to ..)` or `(chdir ..)` part.
It would be used like this: ``pp_with_outputs_to `Target (pp_chdir dir.dir_name (pp_system ("!" :: "ocaml-mdx" :: "pp" :: options_as_list @ [test_file]))) fmt``
Indentation is handled by boxes so it's no longer a concern.

What do you think ?

Reply:
I agree having to care about indentation isn't great and I wouldn't do it like that if it was part of the core of mdx but I'm willing to sacrifice that here, especially given that it makes it more readable than with `pp` combinators. I.e. it's easier to read the plain dune field as a string, even though you proposal uses self-explanatory names.

Eventually I'd prefer piping it into `dune format-dune-file` so that the script code doesn't have to care about formatting at all.

Would that work for you?

Reply:
I don't agree with your first point. I think `pp`s will be more readable, reduce boilerplate, avoid confusing printf and disallow invalid outputs (because they are written once and combinations are restricted to what the API allows).
Also, I think the `pp` approach is easier to review because it's not strings and avoid confusions with format strings syntax.

For example, on that line, the closing `"` is missing.

I wasn't making a point on indentation, but will still be a burden and can be avoided. I think indenting ocaml code (applications of `pp`s) is easier than indenting a string stripped of its context.

Reply:
I think the current implementation is fine enough and a huge improvement over the previous situation.

We can refactor that later but I'd like to get these improvement merged and try to release again.

Reply:
Yes, I agree.